uqcsbot/scripts/events.py

```diff
@@ -0,0 +1,120 @@
+            filter_str = match.group(0)
+            if filter_str in ['full', 'all']:
+                return EventFilter(full=True)
+            elif 'week' in filter_str:
+                return EventFilter(weeks=int(filter_str.split()[0]))
+            else:
+                return EventFilter(cap=int(filter_str))
+
+    def filter_events(self, events: List['Event'], start_time: datetime):
+        if self._weeks is not None:
+            end_time = start_time + timedelta(weeks=self._weeks)
+            return [e for e in events if e.start < end_time]
+        if self._cap is not None:
+            return events[:self._cap]
+
+    def get_header(self):
+        if self._full:
+            return "List of *all* upcoming events"
+        elif self._weeks is not None:
+            return f"Events in the *next _{self._weeks}_ weeks*"
+        else:
+            return f"The *next _{self._cap}_ events*"
+
+    def get_no_result_msg(self):
+        if self._weeks is not None:
+            return f"There doesn't appear to be any events in the next *{self._weeks}* weeks"
+        else:
+            return "There doesn't appear to be any upcoming events..."
+
+
+class Event(object):
+    def __init__(self, start: datetime, end: datetime, location: vText, summary: vText):
+        self.start = start
+        self.end = end
+        self.location = location
+        self.summary = summary
+
+    @staticmethod
+    def from_cal_event(cal_event):
+        start = cal_event.get('dtstart').dt
+        end = cal_event.get('dtend').dt
+        location = cal_event.get('location', 'TBA')
+        summary = cal_event.get('summary')
+        return Event(start, end, location, summary)
+
+    def __str__(self):
+        d1 = self.start.astimezone(BRISBANE_TZ)
+        d2 = self.end.astimezone(BRISBANE_TZ)
+
+        start_str = f"{MONTHS[d1.month-1]} {d1.day} {d1.hour}:{d1.minute:02}"
+        if (d1.month, d1.day) != (d2.month, d2.day):
+            end_str = f"{MONTHS[d2.month-1]} {d2.day} {d2.hour}:{d2.minute:02}"
+        else:
+            end_str = f"{d2.hour}:{d2.minute:02}"
+
+        return f"*{start_str} - {end_str}* - _{self.location}_: `{self.summary}`"
+
+
+@bot.on_command('events')
+async def handle_events(command: Command):
+    event_filter = EventFilter.from_command(command)
+    if not event_filter.is_valid:
+        bot.post_message(command.channel, "Invalid events filter.")
+        return
+
+    http_response = await bot.run_async(get, CALENDAR_URL)
+    cal = Calendar.from_ical(http_response.content)
+
+    current_time = datetime.now(tz=BRISBANE_TZ).astimezone(utc)
+
+    # TODO: support recurring events
+    # subcomponents are how icalendar returns the list of things in the calendar
+    # we are only interested in ones with the name VEVENT as they are events
+    # we also currently filter out recurring events
+    events = [Event.from_cal_event(c) for c in cal.subcomponents if c.name == 'VEVENT' and c.get('RRULE') is None]
```

Comment:
I don't really like the list comprehensions here given they're creating all these intermediate lists, but also it's not enough of a perf worry to bother changing it